### PR TITLE
fix(assets): support exceptions to exclude patterns

### LINKED ISSUE #4450: ecr-assets: incomplete .dockerignore support
Introduced by #4104.

CDK does not handle the common "exclude everything except" pattern for .dockerignore files:

```yaml
# Ignore everything
*
# Allow files and directories
!/src/**
```

Quote from [Docker docs](https://docs.docker.com/engine/reference/builder/#dockerignore-file):

> ... you may want to specify which files to include in the context, rather than which to exclude. To achieve this, specify * as the first pattern, followed by one or more ! exception patterns.

With the current implementation the first 'star' glob effectively excludes everything from the asset folder dyring synth stage. Consequently, deploy stage fails to build the docker image.

### Reproduction Steps

1. Create a ```.dockerignore``` file next to you ```Dockerfile``` with * pattern, like this:
   ```yaml
   # Ignore everything
   *
   # Allow files and directories
   !/src/**
   ```
1. Run ```cdk deploy```

### Error Log

```
unable to prepare context: unable to evaluate symlinks in Dockerfile path: CreateFile C:\Dev\GitHub\playground\EcsWindowsCluster\src\Deployment\cdk.out\asset.1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0\Dockerfile: The system cannot find the file specified.

 ❌  win failed: Error: docker build --tag 305168429857.dkr.ecr.eu-west-1.amazonaws.com/cdk/win:latest cdk.out\asset.1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0 exited with error code 1
docker build --tag 305168429857.dkr.ecr.eu-west-1.amazonaws.com/cdk/win:latest cdk.out\asset.1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0 exited with error code 1
```

### Environment

  - **CLI Version      :** 1.12.0 (build 923055e)
  - **OS               :** Windows

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by nmussy**
Allows adding back files and directories from a previously excluded directory. Previous implementations of `copyDirectory` and `fingerprint` stopped walking through the directory if it was ignored ([see comment in #4450](https://github.com/aws/aws-cdk/issues/4450#issuecomment-541005718))
* expand `dir` exclusions into `dir/*`, to allow `minimist` to match files in folder
* add `listFilesRecursively` method, removing duped code from `copy` and `fingerprint`
* add additional `.dockerignore` test cases into `@aws-cdk/aws-ecr-assets`
* refactor existing test cases

Fixes #4450

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/assets/lib/fs/copy.ts [hunk 1]
```diff
@@ -2,56 +2,45 @@ import fs = require('fs');
...
-      stat = undefined;
-    }
+    {
+      const destFileDir = path.dirname(destFilePath);
 
-    if (stat && stat.isFile()) {
-      fs.copyFileSync(sourceFilePath, destFilePath);
-      stat = undefined;
+      if (!fs.existsSync(destFileDir)) {
+        // FIXME fs.mkdirpSync(destFileDir);
```
Comment (nmussy) [inline packages/@aws-cdk/assets/lib/fs/copy.ts]: I've tried adding `fs-extra` as a depdency to `assets`:

```
lerna add fs-extra@^8.1.0 --scope=@aws-cdk/assets
lerna add --dev @types/fs-extra@^8.0.0 --scope=@aws-cdk/assets
```

But I get the following error when I try building the package:
`Error: Error: Unable to locate module: fs-extra`

Any idea?

Comment (rix0rrr) [reply packages/@aws-cdk/assets/lib/fs/copy.ts]: JSII modules cannot have non-JSII dependencies. You may only depend on JSII modules and the standard Node runtime.

Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/copy.ts]: What should I do then? Add `mkdirpSync` to `@aws-cdk/core`?

Comment (rix0rrr) [reply packages/@aws-cdk/assets/lib/fs/copy.ts]: Reimplementing is the only viable solution.

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 2]
```diff
@@ -58,3 +59,79 @@ export function shouldFollow(mode: FollowMode, sourceRoot: string, realPath: str
...
+      }
+    }
+
+    const relativeFilePath = path.relative(rootDir, fullFilePath);
+
+    // we've just discovered that we have a directory
+    if (stat.isDirectory()) {
+      // to help future shouldExclude calls, we're changing the exlusion patterns
+      // by expliciting "dir" exclusions to "dir/*" (same with "!dir" -> "!dir/*")
+      exclude = exclude.reduce<string[]>((res, pattern) => {
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/lib/fs/utils.ts]: I don't think this should be done in-line while traversing. It is also possible to run through the list of entries in the ignore list, see which ones are directories and treat those as prefixes.

Comment (rix0rrr) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: In fact, I'd like to convert the "list of exclude/include entries" into a function or class that encapsulates the behavior as soon as possible and pass that around.

I'm thinking something like:

```
const ignoreList = parseIgnoreList(rootDirectory, ...list of strings);
```

`ignoreList` will be either of type `(path: string) => boolean`, or be a class with a method of that type.

I would have that as an argument to `listFilesRecursively()`.

File: packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts [hunk 3]
```diff
@@ -80,7 +80,12 @@ export class DockerImageAsset extends Construct implements assets.IAsset {
     const ignore = path.join(dir, '.dockerignore');
 
     if (fs.existsSync(ignore)) {
-      exclude = [...exclude, ...fs.readFileSync(ignore).toString().split('\n').filter(e => !!e)];
+      exclude = [
+        ...exclude,
+        ...fs.readFileSync(ignore).toString().split('\n').filter(e => !!e),
+        // prevents accidentally excluding Dockerfile with a "*"
+        '!Dockerfile',
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts]: Ha, nice.

File: packages/@aws-cdk/aws-ecr-assets/test/dockerignore-image-advanced/.dockerignore [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# This a comment, followed by an empty line
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecr-assets/test/dockerignore-image-advanced/.dockerignore]: Can we do these tests in memory?

I'm thinking of a test where we have preloaded this list of ignore patterns, and pushing a number of paths past it (either by means of `shouldExclude()` or by means of a new `IgnoreList` API)

That is effectively the same and easier to change/assert against than making a whole new fixture with "just these" files in it.

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 5]
```diff
@@ -58,3 +59,79 @@ export function shouldFollow(mode: FollowMode, sourceRoot: string, realPath: str
     return path.resolve(realPath).startsWith(path.resolve(sourceRoot));
   }
 }
+
+export function listFilesRecursively(dir: string, options: CopyOptions & Required<Pick<CopyOptions, 'follow'>>, rootDir?: string): string[] {
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/lib/fs/utils.ts]: Why not make it easy on consumers and instead of returning a list of strings, return a list of objects with metadata?

```ts
interface AssetFile { 
  relativePath: string;
  absolutePath: string;
  isSymlink: boolean;
  isDirectory: boolean;
  symlinkTarget?: string;
}

 export function listFilesRecursively(...): AssetFile[];
```

Avoids having consumers have to do all that `stat`ing again.

In fact you could even do:

```ts
type AssetFile = { 
  relativePath: string;
  absolutePath: string;
  isDirectory: boolean;
} & ({ isSymlink: false } | { isSymlink: true; symlinkTarget: string});
```

To benefit from that sweet sweet TypeScript type checking.

File: packages/@aws-cdk/assets/lib/fs/copy.ts [hunk 6]
```diff
@@ -2,56 +2,44 @@ import fs = require('fs');
...
-      copyDirectory(sourceFilePath, destFilePath, options, rootDir);
-      stat = undefined;
-    }
+    {
+      const destFileDir = path.dirname(destFilePath);
 
-    if (stat && stat.isFile()) {
-      fs.copyFileSync(sourceFilePath, destFilePath);
-      stat = undefined;
+      if (!fs.existsSync(destFileDir)) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/lib/fs/copy.ts]: `mkdirpSync()` makes it so that you don't have to have this initial `if`. Fewer ifs are better!

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 7]
```diff
@@ -58,3 +59,79 @@ export function shouldFollow(mode: FollowMode, sourceRoot: string, realPath: str
...
+
+    const isExcluded = shouldExclude(exclude, relativeFilePath);
+    if (isExcluded && !stat.isDirectory()) {
+      continue;
+    }
+
+    if (stat.isFile()) {
+      files.push(fullFilePath);
+    } else if (stat.isDirectory()) {
+      files.push(...listFilesRecursively(fullFilePath, { ...options, exclude }, rootDir));
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/lib/fs/utils.ts]: For efficiency (no array copying) and not overly passing the same arguments over and over again, I usually prefer the pattern of using an inner function:

```ts
function functionThatObviouslyDoesRecursiveWork(startingPoint, options) {
  const accumulator = [];
  recurse(startingPoint);
  return accumulator;

  function recurse(current) {
    if (condition(current) && options.thing) {
      accumulator.push(current.attribute);
    }
   
     for (const inner of current.children) {
      recurse(inner);
    }
  }
}
```

File: packages/@aws-cdk/assets/lib/fs/utils.ts (hunk 2 above)
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: The issue is that it wouldn't be very economical, `fs` calls-wise. I'm filling the `exclude` directory on the fly because I'm also checking the `stat` of the files I'm traversing.

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 8]
```diff
@@ -58,3 +59,79 @@ export function shouldFollow(mode: FollowMode, sourceRoot: string, realPath: str
     return path.resolve(realPath).startsWith(path.resolve(sourceRoot));
   }
 }
+
+export function listFilesRecursively(dir: string, options: CopyOptions & Required<Pick<CopyOptions, 'follow'>>, rootDir?: string): string[] {
+  const files = [];
+
+  let exclude = [...(options.exclude || [])];
```
[the remark this answers is not in the document]
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: Done

File: packages/@aws-cdk/assets/lib/fs/utils.ts (hunk 5 above)
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: Fun type 👍 
And done

File: packages/@aws-cdk/assets/lib/fs/copy.ts (hunk 6 above)
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/copy.ts]: Done

File: packages/@aws-cdk/assets/lib/fs/utils.ts (hunk 7 above)
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: Done

**Comment by nmussy** [comment]
@rix0rrr Your new PR title isn't completely accurate
* This PR affects all `exclude` properties (including `lambda.Code.fromAsset`), not just `.dockerignore`
* Maybe "support exclusion exceptions" would make more sense, as "exceptions" previously existed?

**Comment by rix0rrr** [comment]
> @rix0rrr Your new PR title isn't completely accurate
> 
> * This PR affects all `exclude` properties (including `lambda.Code.fromAsset`), not just `.dockerignore`
> * Maybe "support exclusion exceptions" would make more sense, as "exceptions" previously existed?

Yeah, sounds good.

File: packages/@aws-cdk/aws-ecr-assets/test/dockerignore-image-advanced/.dockerignore (hunk 4 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-ecr-assets/test/dockerignore-image-advanced/.dockerignore]: I ended up creating a `createFsStructureFromTree` function, and removing the previous fixtures (with the exception of `demo-image/`).

Does that work for you?

File: packages/@aws-cdk/assets/lib/fs/fingerprint.ts [hunk 9]
```diff
@@ -38,40 +38,28 @@ export function fingerprint(fileOrDirectory: string, options: FingerprintOptions
...
-
-  function _processFileOrDirectory(symbolicPath: string, realPath = symbolicPath) {
-    if (shouldExclude(exclude, symbolicPath)) {
-      return;
-    }
-
-    const stat = fs.lstatSync(realPath);
-    const relativePath = path.relative(fileOrDirectory, symbolicPath);
+  for (const assetFile of listFilesRecursively(fileOrDirectory, {...options, follow}, rootDirectory)) {
+    const relativePath = path.relative(fileOrDirectory, assetFile.absolutePath);
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/lib/fs/fingerprint.ts]: Can you not use `relativePath` from the AssetFile here?

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 10]
```diff
@@ -58,3 +59,112 @@ export function shouldFollow(mode: FollowMode, sourceRoot: string, realPath: str
...
+
+      const relativeFilePath = path.relative(rootDir, fullFilePath);
+
+      // we've just discovered that we have a directory
+      if (stat.isDirectory()) {
+        // to help future shouldExclude calls, we're changing the exlusion patterns
+        // by expliciting "dir" exclusions to "dir/*" (same with "!dir" -> "!dir/*")
+        exclude = exclude.reduce<string[]>((res, pattern) => {
+          res.push(pattern);
+          if (pattern.trim().replace(/^!/, '') === relativeFilePath) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/lib/fs/utils.ts]: Does not look like this will correctly handle cases where the directory entry in the ignore list contains a wildcard.

File: packages/@aws-cdk/aws-ecr-assets/test/test.image-asset.ts [hunk 11]
```diff
@@ -212,19 +213,38 @@ export = {
     const app = new App();
     const stack = new Stack(app, 'stack');
 
-    const image = new DockerImageAsset(stack, 'MyAsset', {
-      directory: path.join(__dirname, 'dockerignore-image')
-    });
+    const directory = fs.mkdtempSync(path.join(os.tmpdir(), 'dockerignore-image'));
+    createFsStructureFromTree(directory, `
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecr-assets/test/test.image-asset.ts]: Oh that is clever and pretty sexy!

What I was thinking when I made my previous remark though is testing the exclude pattern evaluator separately from the `listFilesRecursively` operation. So I was thinking something like:

```ts
const shouldInclude = parseExcludePattern(['foobar.txt']); 
// or whatever API we choose, doesn't have to directly return a function, maybe it is a class but
// it is something that should give me a yes/no answer

test.ok(shouldInclude('Dockerfile'));
test.ok(shouldInclude('subdirectory/baz.txt'));
test.ok(!shouldInclude('foobar.txt'));
test.ok(!shouldInclude('subdirectory/foobar.txt'));
```

I can then test various edge case behaviors of the exclude pattern evaluator in-memory (speaking of `fs`-calls efficiency!) and then I only have to have 1 or 2 tests, maybe they're already the integ tests, to make sure that `listFilesRecursively` uses the exclude pattern evaluator in the right way.

**Comment by rix0rrr** [comment]
If I'm missing something and it's not that simple as I'm thinking it is, can you please enlighten me with some edge cases? :)

File: packages/@aws-cdk/assets/lib/fs/fingerprint.ts (hunk 9 above)
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/fingerprint.ts]: It is relative to `fileOrDirectory` instead of `rootDirectory`

**Comment by nmussy** [comment]
There's a slight issue with your ideal high-level code, which amounts to the state of the agorithm pre-PR:
```ts
function recursiveList(current, shouldInclude) {
  if (shouldInclude(current)) {
    if (isFile(current)) {
      result.push(current);
    }
    recurseIntoChildren(current);
  } else {
    // current might be ignored, but it's children are included
  }
}
```

As for this comment:

> And in fact, I think the evaluator can be written to not need to make any stat calls itself at all.

Again, I need to check that the excluded/included pattern is a directory, to see if I need to walk through it or not.

I might be the one missing something here though, sorry if that's the case

File: packages/@aws-cdk/assets/lib/fs/utils.ts (hunk 10 above)
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: Can you give me an example? Because I think I'm already testing this use-case:
https://github.com/aws/aws-cdk/blob/baa431539a0b024d9f065cfed8a27e24771ce6cc/packages/%40aws-cdk/aws-ecr-assets/test/test.image-asset.ts#L339

File: packages/@aws-cdk/aws-ecr-assets/test/test.image-asset.ts (hunk 11 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-ecr-assets/test/test.image-asset.ts]: This will only be possible if we resolve the main-thread discussion about needing the `fs` to evaluate the exclusion list:

```ts
const shouldInclude = parseExcludePattern(['dir', '!dir/file']); 

// naively, this would be false
// but we would need to create dir if `dir/file` exists
test.ok(shouldInclude('dir'));

test.ok(shouldInclude('dir/file'));
test.ok(!shouldInclude('dir/other_file'));
```

**Comment by rix0rrr** [comment]
> There's a slight issue with your ideal high-level code, which amounts to the state of the algorithm pre-PR:

Depends on when `shouldInclude()` returns true or false. I think I agree with you that we need to know whether the current thing is a directory or not. I don't agree with us needing to modify the exclude list while iterating over it.

Effectively, we want to make it do this:

```ts
function shouldInclude(path, isDirectory) {
  return !inExcludeList(path) || (isDirectory && anyChildrenExcepted(path));
}
```

So then the question is, how do we implement `anyChildrenExcepted`. 

...
...
...
... 

After thinking on it some more, depending on the interpretation rules of `.dockerignore`, I think we just cannot eagerly skip directories at all. Is the following supposed to work?

```
# File tree
      ├── deep
      │   ├── dir
      │   │   └── struct
      │   │       └── qux.txt
      

# .dockerignore
deep
!struct
```

Should this include `qux.txt` or not? If this is supposed to include `qux.txt` then there is no way to write `anyChildrenExcepted` without knowing the entire file tree, at which point there is no advantage to doing the filtering in-line (might as well have 2 cleanly separated phases).

If this is NOT supposed to include `qux.txt` because you're supposed to write it like this:

```
deep
!deep/dir/struct
```

Then there IS a way to write `anyChildrenExcepted` going completely from memory, given the input path `deep`. The rules would be:

* Line starts with a `!`
* Part of the line up to (and excluding) a `/` character matches the current path as a glob (i.e., `deep` or `d??p` or `d*`).
* There is at least one character after the `/`

**Comment by nmussy** [comment]
> Is the following supposed to work?
> ```
> deep
> !struct
> ```

Not that I'm aware of, but I'll double-check.

---

I see where you're going with `anyChildrenExcepted`, this might work.

The main I issue that I see with going out of memory only is the theoretical inclusion of empty directories:

```
# File tree
      ├── deep
      │   ├── dir
      │   │   └── struct
      │   │       └── qux.txt
      

# .dockerignore
deep
!deep/dir/struct/foo.txt
```

Doing this in memory only would mean that `anyChildrenExcepted` would be true for `deep`, `dir` and `struct`. We would still need to check during the execution of `copyDirectory` that `foo.txt` exists (which it doesn't in my example.)

**Comment by nmussy** [comment]
> > Is the following supposed to work?
> > ```
> > deep
> > !struct
> > ```
> 
> Not that I'm aware of, but I'll double-check.

Confirmed, `!deep/dir/struct` needs to be explicitly set

**Comment by rix0rrr** [comment]
> We would still need to check during the execution of `copyDirectory` that `foo.txt` exists (which it doesn't in my example.)

Sure. What that means is we had a potential for an optimization, and we didn't take it. That is not so bad. We did about 2 `stat` calls more than strictly necessary. At least we're not incorrect!

And note that even the following directory struture:

```
      ├── deep
      │   ├── dir
      │   ├── bigdir
      │   │   └── 1 MILLION FILES HERE

# And the ignore pattern
deep
!deep/dir/doesnotexist.txt
```

Isn't really that horrible. We recurse into `deep`, then into `dir`, and match nothing. Crucially, we don't recurse into `bigdir`!

It's only really bad if we have a directory with lots of entries (let's say, millions) out of which only one file is excepted, and that file doesn't even exist (in this case, `!deep/bigdir/doesnotexist.txt`). We would then proceed to list and try to match a million files for nothing. How likely is that to even happen?

And in fact, no amount of coding around an exception list would solve that. The only way to solve that is to move to an *inclusion* list instead of an *exclusion* list.

So all in all, I don't think it's a big risk.

**Comment by nmussy** [comment]
Yeah, you make a good point.

I'll start working on it tomorrow. I'll get back to you if I stumble on anything big, but I think your algorithm is solid (not to mention cleaner, which is definitely no argument).

**Comment by nmussy** [comment]
@rix0rrr I've added a `shouldExcludeDeep` method and tested it in https://github.com/aws/aws-cdk/pull/4473/commits/273d3f51904825e33e96b0d38662644f0d09b7cb. I haven't updated `listFilesRecursively` yet, but it's a promising start! 👍

**Comment by nmussy** [comment]
@rix0rrr That was pretty fun (and slightly painful). It should be good for review 👍

**Comment by rix0rrr** [comment]
> @rix0rrr That was pretty fun (and slightly painful). It should be good for review 👍

Thanks for putting up with me :)

**Comment by nmussy** [comment]
You had a clearer idea than I did, and the PR is much better for it. Thanks for having the patience of convincing me 😄

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 12]
```diff
@@ -1,34 +1,113 @@
...
+/**
+ * Determines whether a given file should be excluded,taking into account deep file structures
+ *
+ * @param exclude  exclusion patterns
+ * @param filePath file path to be assessed against the pattern
+ */
+export function shouldExcludeDeep(exclude: string[], relativePath: string): boolean {
+  const [_shouldExclude] = relativePath.split(path.sep).reduce<[boolean, number, string]>(
+    ([accExclude, accPriority, pathIterator], pathComponent) => {
+      pathIterator = path.join(pathIterator, pathComponent);
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/lib/fs/utils.ts]: I feel it would be cleaner if you made a function like `pathComponents()` (which given a path like `a/b/c` would return `[a, a/b, a/b/c]`) and iterated over that.

Again I feel a simple `for` loop would be more natural. If you really want to make it functional, why not something like:

```ts
const resultsAndPrios = pathComponents(relativePath).map(pc => shouldExcludePriority(rules,  pc));
const [shouldExclude] = pickHighest(resultsAndPrios, x => x[1]);
return shouldExclude;
```

Requires you to write some helpers, but to me is a lot clearer what is happening.

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 13]
```diff
@@ -1,34 +1,113 @@
...
+
+      return [accExclude, accPriority, pathIterator];
+    }, [false, -1, '']);
+
+  return _shouldExclude;
+}
+
+/**
+ * Determines whether a given directory should be excluded and not explored further
+ * This might be true even if the directory is explicitly excluded,
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/lib/fs/utils.ts]: In the case you're describing, shouldn't the function return `false` rather than `true?

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 14]
```diff
@@ -1,34 +1,113 @@
...
+/**
+ * Determines whether a given file should be excluded or not based on given
+ * exclusion glob patterns.
+ *
+ * @param exclude  exclusion patterns
+ * @param filePath file patg to be assessed against the pattern
+ *
+ * @returns `true` if the file should be excluded, followed by the index of the rule applied
+ */
+export function shouldExcludePriority(exclude: string[], filePath: string): [boolean, number] {
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: The name confuses me. Proposed name  change: how about `evaluateExcludePatterns`.

Or in fact, I've been saying this a couple of times already: please put these related functions that operate on the same data structure in a class. They're not as independent as you would like.

```ts
class ExcludeRules {
  constructor(private readonly patterns: string[]) { }

  public excludeFile(filePath: string): boolean { ... }
  public excludeDirectory(dirPath: string): boolean { ... }
  
  private evaluate(pathName: string): [boolean, number] { ... }

}
```

File: packages/@aws-cdk/assets/test/fs/test.utils.ts [hunk 15]
```diff
@@ -192,4 +194,329 @@ export = {
...
+      ], [
+        'bar.txt',
+        'foo',
+        'foo.txt.old',
+      ]);
+
+      test.done();
+    },
+    'contridactory'(test: Test) {
+      testShouldExcludeDeep(test, ['foo.txt', '!foo.txt'], [], ['foo.txt']);
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/test/fs/test.utils.ts]: How about contradictory with the order reversed?

File: packages/@aws-cdk/assets/test/fs/test.utils.ts [hunk 16]
```diff
@@ -192,4 +194,329 @@ export = {
       },
     }
   },
+
+  shouldExcludeDeep: {
```
Comment (rix0rrr) [inline packages/@aws-cdk/assets/test/fs/test.utils.ts]: Thanks for a thorough test suite!

File: packages/@aws-cdk/assert/lib/fs-utils.ts [hunk 17]
```diff
@@ -0,0 +1,98 @@
+import fs = require('fs');
```
[the remark this answers is not in the document]
Comment (nmussy) [reply packages/@aws-cdk/assert/lib/fs-utils.ts]: I moved it to `assert` because I was using it both in `assets` and `aws-ecr-assets`.  I could move it to `assets` instead, and expose it to make to available to `aws-ecr-assets`?

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 18]
```diff
@@ -1,34 +1,113 @@
...
+  return exclude.reduce<[boolean, number]>((res, pattern, patternIndex) => {
```
[the remark this answers is not in the document]
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: I was originally using a `map` to have both `pattern` and `patternIndex`. I'll revert back to it.

File: packages/@aws-cdk/assert/lib/fs-utils.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/assert/lib/fs-utils.ts]: I see. Yes, I think put it in `assets`, and in fact you could import it directly from the `test/` directory in `aws-ecr-assets`:

```ts
import fsUtils = require('@aws-cdk/assets/test/fs-utils');
```

It's not super nice, but I don't want test helpers to be part of the public API of a package. This usage would be okay by me since it's between 1st-party packages only. Would be nice to spend a couple of comment lines on it when you're making that change.

File: packages/@aws-cdk/assets/test/fs/test.utils.ts (hunk 15 above)
Comment (nmussy) [reply packages/@aws-cdk/assets/test/fs/test.utils.ts]: Done

File: packages/@aws-cdk/assets/lib/fs/utils.ts (hunk 13 above)
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: Done

File: packages/@aws-cdk/assets/lib/fs/utils.ts (hunk 12 above)
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: I did some refactoring, preferring `for` loops over `reduce`s. I've only kept `map`s when I needed to know both the current array value and index.

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 19]
```diff
@@ -1,34 +1,113 @@
...
+
+/**
+ * Determines whether a given directory should be excluded and not explored further
+ * This might be true even if the directory is explicitly excluded,
+ * but one of its children might be inclunded
+ *
+ * @param exclude  exclusion patterns
+ * @param directoryPath directory path to be assessed against the pattern
+ */
+export function shouldExcludeDirectory(exclude: string[], directoryPath: string): boolean {
```
[the remark this answers is not in the document]
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: I was exporting it to test it. It's the same now with the public method

File: packages/@aws-cdk/assets/lib/fs/utils.ts [hunk 20]
```diff
@@ -1,34 +1,113 @@
...
+  }, [false, -1]);
+}
+
+/**
+ * Determines whether a given file should be excluded,taking into account deep file structures
+ *
+ * @param exclude  exclusion patterns
+ * @param filePath file path to be assessed against the pattern
+ */
+export function shouldExcludeDeep(exclude: string[], relativePath: string): boolean {
```
[the remark this answers is not in the document]
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: I was exporting it to test it. It's the same now with the public method

File: packages/@aws-cdk/assert/lib/fs-utils.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (nmussy) [reply packages/@aws-cdk/assert/lib/fs-utils.ts]: Done

File: packages/@aws-cdk/assets/lib/fs/utils.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: Done

File: packages/@aws-cdk/assets/lib/fs/utils.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (nmussy) [reply packages/@aws-cdk/assets/lib/fs/utils.ts]: I've switched to a class, that I'm still exporting to be able to test it. Same for your `evaluate`/my `evaluateFile`, it needs to be public.

**Comment by nmussy** [comment]
Thanks again @rix0rrr 🎉
